src/hotspot/share/gc/shenandoah/shenandoahAgeCensus.cpp

```diff
@@ -0,0 +1,383 @@
+      CENSUS_NOISE(_global_noise[_epoch].merge(_local_noise[i]);)
+      CENSUS_NOISE(_local_noise[i].clear();)
+    }
+  } else {
+    // census during evac
+    assert(pv1 != nullptr && pv2 != nullptr, "Error, check caller");
+    _global_age_table[_epoch]->merge(pv1);
+    _global_age_table[_epoch]->merge(pv2);
+  }
+
+  update_tenuring_threshold();
+
+  // used for checking reasonableness of census coverage, non-product
+  // only.
+  NOT_PRODUCT(update_total();)
+}
+
+
+// Reset the epoch for the global age tables,
+// clearing all history.
+void ShenandoahAgeCensus::reset_global() {
+  assert(_epoch < MAX_SNAPSHOTS, "Out of bounds");
+  for (uint i = 0; i < MAX_SNAPSHOTS; i++) {
+    _global_age_table[i]->clear();
+    CENSUS_NOISE(_global_noise[i].clear();)
+  }
+  _epoch = MAX_SNAPSHOTS;
+  assert(_epoch < MAX_SNAPSHOTS, "Error");
+}
+
+// Reset the local age tables, clearing any partial census.
+void ShenandoahAgeCensus::reset_local() {
+  if (!ShenandoahGenerationalAdaptiveTenuring || ShenandoahGenerationalCensusAtEvac) {
+    assert(_local_age_table == nullptr, "Error");
+    return;
+  }
+  size_t max_workers = ShenandoahHeap::heap()->max_workers();
+  for (uint i = 0; i < max_workers; i++) {
+    _local_age_table[i]->clear();
+    CENSUS_NOISE(_local_noise[i].clear();)
+  }
+}
+
+#ifndef PRODUCT
+// Is global census information clear?
+bool ShenandoahAgeCensus::is_clear_global() {
+  assert(_epoch < MAX_SNAPSHOTS, "Out of bounds");
+  for (uint i = 0; i < MAX_SNAPSHOTS; i++) {
+    bool clear = _global_age_table[i]->is_clear();
+    CENSUS_NOISE(clear |= _global_noise[i].is_clear();)
+    if (!clear) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Is local census information clear?
+bool ShenandoahAgeCensus::is_clear_local() {
+  if (!ShenandoahGenerationalAdaptiveTenuring || ShenandoahGenerationalCensusAtEvac) {
+    assert(_local_age_table == nullptr, "Error");
+    return true;
+  }
+  size_t max_workers = ShenandoahHeap::heap()->max_workers();
+  for (uint i = 0; i < max_workers; i++) {
+    bool clear = _local_age_table[i]->is_clear();
+    CENSUS_NOISE(clear |= _local_noise[i].is_clear();)
+    if (!clear) {
+      return false;
+    }
+  }
+  return true;
+}
+
+size_t ShenandoahAgeCensus::get_all_ages(uint snap) {
+  assert(snap < MAX_SNAPSHOTS, "Out of bounds");
+  size_t pop = 0;
+  const AgeTable* pv = _global_age_table[snap];
+  for (uint i = 0; i < MAX_COHORTS; i++) {
+    pop += pv->sizes[i];
+  }
+  return pop;
+}
+
+size_t ShenandoahAgeCensus::get_skipped(uint snap) {
+  assert(snap < MAX_SNAPSHOTS, "Out of bounds");
+  return _global_noise[snap].skipped;
+}
+
+void ShenandoahAgeCensus::update_total() {
+  _counted = get_all_ages(_epoch);
+  _skipped = get_skipped(_epoch);
+  _total   = _counted + _skipped;
+}
+#endif // !PRODUCT
+
+void ShenandoahAgeCensus::update_tenuring_threshold() {
+  if (!ShenandoahGenerationalAdaptiveTenuring) {
+    _tenuring_threshold[_epoch] = InitialTenuringThreshold;
```

Comment:
`InitialTenuringThreshold` is currently only used by parallel GC, so that that flag constraint is only implemented in parallel GC [here](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp#L177) . Now, Shenandoah uses the flag, you may want to consider to parallel GC's implementation into `shared`.

Reply:
https://bugs.openjdk.org/browse/JDK-8343226